rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,418 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+
+# RFC-63: Functional Indexes
+
+## Proposers
+
+- @yihua
+- @alexeykudinkin
+- @codope
+
+## Approvers
+
+- @vinothchandar
+- @xushiyan
+- @nsivabalan
+
+## Status
+
+JIRA: [HUDI-512](https://issues.apache.org/jira/browse/HUDI-512)
+
+## Abstract
+
+In this RFC, we propose **Functional Indexes**, a new capability to
+Hudi's [multi-modal indexing](https://hudi.apache.org/blog/2022/05/17/Introducing-Multi-Modal-Index-for-the-Lakehouse-in-Apache-Hudi)
+subsystem that offers a compelling vision to support not only accelerating queries but also reshape partitions as
+another layer of the indexing system, abstracting them from the traditional fixed notion, while providing flexibility
+and performance.
+
+## Background
+
+Hudi employs multi-modal indexing to optimize query performance. These indexes, ranging from simple files index to
+record-level indexing, cater to a diverse set of use cases, enabling efficient point lookups and reducing the data
+scanned during queries. This is usually done in two ways:
+
+- **Partition pruning**:  The partition pruning relies on a table with physical partitioning, such as Hive partitioning.
+  A partitioned table uses a chosen column such as the date of `timestamp` and stores the rows with the same date to the
+  files under the same folder or physical partition, such as `date=2022-10-01/`. When the predicate in a query
+  references the partition column of the physical partitioning, the files in the partitions not matching the predicate
+  are filtered out, without scanning. For example, for the predicate `date between '2022-10-01' and '2022-10-02'`, the
+  partition pruning only returns the files from two partitions, `2022-10-01` and `2022-10-02`, for further processing.
+  The granularity of the pruning is at the partition level.
+
+
+- **Data Skipping**:  Skipping data at the file level, with the help of column stats or record-level index. For example,
+  with column stats index containing minimum and maximum values of a column for each file, the files falling out of the
+  range of the values compared to the predicate can be pruned. For a predicate with `age < 20`, the file pruning filters
+  out a file with columns stats of `[30, 40]` as the minimum and maximum values of the column `age`.
+
+While Hudi already supports partition pruning and data skipping for different query engines, we
+recognize that the following use cases need better query performance and usability:
+
+- Data skipping based on functions defined on column(s)
+- Support for different storage layouts and view partition as index
+- Support for secondary indexes
+
+Next, we explain these use cases in detail.
+
+### Use Case 1: Data skipping based on functions defined on column(s)
+
+Let's consider a non-partitioned table containing the events with a `timestamp` column. The events with naturally
```

Comment:
the column is named "timestamp"? rename to `ts` to disambiguate from the actual datatype?